src/client/Microsoft.Identity.Client/OAuth2/MsalTokenResponse.cs

```diff
@@ -146,6 +146,32 @@ internal static MsalTokenResponse CreateFromiOSBrokerResponse(Dictionary<string,
             return response;
         }
 
+        internal static MsalTokenResponse CreateFromAppProviderResponse(ExternalTokenResult tokenResponse)
+        {
+            var response = new MsalTokenResponse
```

Comment:
We should do some validation of `tokenResponse` - it is mandatory to have most of these params, otherwise the token cache can be left in a bad state.